### PR TITLE
fixes: issue 999 refactor db select to reduce cognitive complexity

### PR DESCRIPTION
**Pull request by vigneshp91**
Fixes https://github.com/gofr-dev/gofr/issues/999
**Description:**

-  To reduce the cognitive complexity created two functions to handle the select operation.

### CODE REVIEW

**Review by aryanmehrotra — APPROVED** [review]

File: pkg/gofr/datasource/sql/db.go [hunk 1]
```diff
@@ -231,37 +230,44 @@ func (d *DB) Select(ctx context.Context, data interface{}, query string, args ..
...
-		}
+	if rvo.Elem().CanSet() {
+		rvo.Elem().Set(rv)
+	}
+}
 
-	default:
-		d.logger.Debugf("a pointer to %v was not expected.", rv.Kind().String())
+func (d *DB) selectStruct(ctx context.Context, query string, args []interface{}, rv reflect.Value) {
+	rows, _ := d.QueryContext(ctx, query, args...)
```
Comment (ccoVeille) [inline pkg/gofr/datasource/sql/db.go:268]: Why ignoring the error?

If it's legitimate, please add a comment

I noted previous code was not doing it, but it looks odd

Comment (ccoVeille) [reply pkg/gofr/datasource/sql/db.go]: Thanks

**Comment by vigneshp91** [comment]
@aryanmehrotra Can you approve and merge this?

File: pkg/gofr/datasource/sql/db.go [hunk 2]
```diff
@@ -231,37 +230,49 @@ func (d *DB) Select(ctx context.Context, data interface{}, query string, args ..
...
+func (d *DB) selectStruct(ctx context.Context, query string, args []interface{}, rv reflect.Value) {
+	rows, err := d.QueryContext(ctx, query, args...)
+	if err != nil {
+		d.logger.Errorf("error running query: %v", err)
+		return
+	}
+
+	for rows.Next() {
+		d.rowsToStruct(rows, rv)
 	}
```
Comment (ccoVeille) [inline pkg/gofr/datasource/sql/db.go:276]: No error management to do with something like this?


```suggestion
	for rows.Next() {
		d.rowsToStruct(rows, rv)
 	}

	if err:= rows.Err() {
		return err
 	}
```

File: pkg/gofr/datasource/sql/db.go [hunk 3]
```diff
@@ -231,37 +230,49 @@ func (d *DB) Select(ctx context.Context, data interface{}, query string, args ..
...
+			_ = rows.Scan(val.Interface())
 		}
 
-	case reflect.Struct:
-		rows, _ := d.QueryContext(ctx, query, args...)
-		for rows.Next() {
-			d.rowsToStruct(rows, rv)
-		}
+		rv = reflect.Append(rv, val.Elem())
+	}
```
Comment (ccoVeille) [inline pkg/gofr/datasource/sql/db.go:260]: Same question as here

https://github.com/gofr-dev/gofr/pull/1019#discussion_r1760304447

**Comment by Umang01-hash** [comment]
@vigneshp91 seems like your PR has some review comments unresolved. Please address them

File: pkg/gofr/datasource/sql/db.go (hunk 2 above)
Comment (Umang01-hash) [reply pkg/gofr/datasource/sql/db.go:276]: I have added a check for `rows.Err()` and if it is not nil returning from there after logging

File: pkg/gofr/datasource/sql/db.go (hunk 3 above)
Comment (Umang01-hash) [reply pkg/gofr/datasource/sql/db.go:260]: Also added a check here

**Review by vipul-rawat — APPROVED** [review]

**Review by aryanmehrotra — APPROVED** [review]

**Pull request merged**